gobblin-compaction/src/main/java/gobblin/compaction/source/CompactionSource.java

```diff
@@ -18,76 +25,296 @@
+            } else {
+              VerifiedDataset vd = ((Either.Left<VerifiedDataset, DatasetVerificationException>) either).getLeft();
+              if (!vd.verifiedResult.isVerified) {
+                log.error ("Dataset {} verification has failed but should retry {}", vd.dataset.datasetURN(),
+                        vd.verifiedResult.shouldRetry);
+                if (vd.verifiedResult.shouldRetry) {
+                  failedDatasets.add(vd.dataset);
+                }
+              }
+            }
+          }
+
+          this.datasets = failedDatasets;
+          if (stopwatch.elapsed(TimeUnit.SECONDS) > timeOut) {
+            break;
+          }
+        }
+
+        if (this.datasets.size() > 0) {
+          for (Dataset dataset: datasets) {
+            log.info ("{} is timed out and give up the verification, adding a failed task", dataset.datasetURN());
+            // create failed task for these failed datasets
+            this.workUnitIterator.addWorkUnit (createWorkUnitForFailure(dataset));
+          }
+        }
+
+        this.workUnitIterator.done();
+      } catch (RuntimeException e) {
+        throw e;
+      } catch (Exception e) {
+        throw new RuntimeException(e);
+      }
+
+    }
+  }
 
+  @AllArgsConstructor
+  private static class DatasetVerificationException extends Exception {
+    private Dataset dataset;
+    private Throwable cause;
 
-      for (Dataset dataset: datasets) {
-        // all verifier should be passed before we compact the dataset
-        boolean verificationPassed = true;
+    public String toString () {
+      return "Exception :" + cause + " from " + dataset.datasetURN();
+    }
+  }
+
+  @AllArgsConstructor
+  private static class VerifiedDataset {
+    private Dataset dataset;
+    private VerifiedResult verifiedResult;
+  }
+
+  @AllArgsConstructor
+  private static class VerifiedResult {
+    private boolean isVerified;
+    private boolean shouldRetry;
+  }
+
+  @AllArgsConstructor
+  private class DatasetVerifier implements Callable {
+    private Dataset dataset;
+    private CompactionWorkUnitIterator workUnitIterator;
+    private List<CompactionVerifier> verifiers;
+
+    /**
+     * {@link VerifiedDataset} wraps original {@link Dataset} because if verification failed, we are able get original
+     * datasets and restart the entire process of verification against those failed datasets.
+     */
+    public VerifiedDataset call () throws DatasetVerificationException {
+      try {
+        VerifiedResult result = this.verify(dataset);
+        if (result.isVerified) {
+          this.workUnitIterator.addWorkUnit (createWorkUnit(dataset));
+        }
+        return new VerifiedDataset(dataset, result);
+      } catch (Exception e) {
+        throw new DatasetVerificationException(dataset, e);
+      }
+    }
+
+    public VerifiedResult verify (Dataset dataset) throws IOException {
+      boolean verificationPassed = true;
+      boolean shouldRetry = false;
```

Comment:
I think it should be the opposite. `shouldRetry` is an `and` of `shouldRetry` of all the individual verifications. Then, `shouldRetry` starts as true, and if any verification sets it to false then it remains false.